### ISSUE TITLE
[Feature Request] Store KV cache of computed prompts to disk to avoid re-compute in follow-up runs

### ISSUE THREAD

**Issue by mark-lord** [ml-explore/mlx-examples#917]
It would be great if MLX_lm supported a --cache_prompt flag like in llama.cpp's integration (discussed in [this issue](https://github.com/ml-explore/mlx-examples/issues/new)). 

This would be a big benefit in reducing latency / start up time for repeated runs that include the same prompts - e.g. chatbot applications with a running chatlog, or when using long multishot examples that take up a lot of tokens (very useful for production environments).

I'm no expert so likely won't be able to assist with a PR on this :'( But looking at the discussion for the initial llama.cpp integration, I have a couple of observations:
- It seems that the cache should store the model's memory_k and memory_v data, along with other necessary state information incl. (but not limited to?) n_past, RNG state, logits, and embedding vectors
- I think the llama.cpp team may have implemented some form of lightweight compression (zstd?) for the cached data, as apparently the memory tensors are often compressible, especially when the full context is not used

The default behaviour of --cache-prompt might be to just save to the model folder like trained adapters; though it'd be useful to then have additional flags to maybe clear or delete the cache afterwards

**Comment by Jckwind** [comment]
take a look at this open source library / repo: https://github.com/otriscon/llm-structured-output they have an implementation of a reusable KV cache for mlx. i've gotten it working - works surprisingly well!

**Comment by awni** [comment]
I think we can close this! 🚀
